src/serac/numerics/assembled_sparse_matrix.cpp

```diff
@@ -0,0 +1,297 @@
+
+#include "mfem/linalg/dtensor.hpp"
+#include "assembled_sparse_matrix.hpp"
+
+namespace detail {
+/**
+ * We'd like to access the protected offset, indices, and gatherMap variables from mfem::ElementRestriction
+ * to compute our sparsity patterns in a manner consistent with MFEM.
+ * In order to do that, we use the following template tricks.
+ * Here's an article that explains the trick being used: https://accu.org/journals/overload/28/156/harrison_2776/
+ */
+template <auto offsetV, auto indicesV, auto gatherV>
+struct forbidden_restriction {
+  friend mfem::Array<int>& ElementRestriction_offsets(mfem::ElementRestriction& from) { return from.*offsetV; }
+  friend mfem::Array<int>& ElementRestriction_indices(mfem::ElementRestriction& from) { return from.*indicesV; }
+  friend mfem::Array<int>& ElementRestriction_gatherMap(mfem::ElementRestriction& from) { return from.*gatherV; }
+};
+
+mfem::Array<int>& ElementRestriction_offsets(mfem::ElementRestriction&);
+mfem::Array<int>& ElementRestriction_indices(mfem::ElementRestriction&);
+mfem::Array<int>& ElementRestriction_gatherMap(mfem::ElementRestriction&);
+
+template struct forbidden_restriction<&mfem::ElementRestriction::offsets, &mfem::ElementRestriction::indices,
+                                      &mfem::ElementRestriction::gatherMap>;
+/**
+ * @brief Account for HCURL-oriented index
+ * @param[in] index Index-value. Can be negative for HCURL
+ * @return corresponding valid index value in array (must be positive)
+ */
+
+template <typename T>
+T oriented_index(T index)
+{
+  return index >= 0 ? index : -1 - index;
+}
+
+/**
+ * @brief Returns sign of index
+ * @param[in] index Index-value.
+ * @return sign of index. +1 for 0 or positive, and -1 for negative
+ */
+
+template <typename T>
+T sign(T index)
+{
+  return index >= 0 ? 1 : -1;
+}
+
+/**
+ * @brief Adds HCURL orientation encoding to index
+ * @param[in] index positive index value
+ * @param[in] orientation Positive orientation or negative
+ */
+template <typename T>
+T orient_index(T index, bool pos)
```

Comment:
Naming, also, it isn't immediately obvious to me what the difference between `orientedIndex` and `orientIndex` would be, is there a different set of names or further clarification that could be provided?